include/envoy/thread_local/thread_local.h

```diff
@@ -228,5 +229,24 @@ class Instance : public SlotAllocator {
   virtual Event::Dispatcher& dispatcher() PURE;
 };
 
+struct MainThread {
+  using MainThreadSingleton = InjectableSingleton<MainThread>;
+  bool inMainThread() { return main_thread_id_ == std::this_thread::get_id(); }
+  static void init() {
+    static MainThread main_thread;
+    main_thread = MainThread();
+    MainThreadSingleton::initialize(&main_thread);
+  }
+  static void clear() { MainThreadSingleton::clear(); }
+  static bool isMainThread() { return MainThreadSingleton::get().inMainThread(); }
+
+private:
+  std::thread::id main_thread_id_{std::this_thread::get_id()};
+};
+
+#define envoy_try                                                                                  \
```

Comment:
I'd remove this and move to the next PR.

Reply:
Will do.